### PR TITLE
[Yul] Activate load resolver.

### LINKED ISSUE #6865: Activate SLoad resolver
Eliminating sload instructions where the value in storage can be known was implemented in https://github.com/ethereum/solidity/pull/6807 but it has not been added to the OptimizerSuite yet.

### LINKED ISSUE #6962: [Yul] Activate LoadResolver
The LoadResolver has been implemented in https://github.com/ethereum/solidity/pull/6807 but not yet activated in the OptimizerSuite. It replaces mload/sload opcodes by the actual value, if it is known.

### PR DESCRIPTION
**Pull request by chriseth**
Fixes https://github.com/ethereum/solidity/issues/6865
Fixes https://github.com/ethereum/solidity/issues/6962

### CODE REVIEW

**Comment by chriseth** [comment]
@bshastry can you please fuzz this? I did not change the trace comparison algorithm because I think that `mload` and `sload` instructions are not logged at all.

**Comment by chriseth** [comment]
I still would like to see if it is easy to tell the component that memory is zero-initialized.

**Comment by bshastry** [comment]
@chriseth The following test input causes a trace divergence between unoptimized and optimized versions

```
{
  function foo() {
    mstore(0x2200000000000004144717016770310599, 1)
  }
  foo()
  function bar()
  {
    mstore(mload(0x2200000000000004144717016770310599), 3)
  }
  bar()
}
```

The optimized code is
```
            mstore(0x2200000000000004144717016770310599, 1)
            mstore(1, 3)
```


The traces are

```
// Pre optimization
Trace:
Memory dump:
Storage dump:

// Post optimization
Trace:
Memory dump:
    20: 0300000000000000000000000000000000000000000000000000000000000000
Storage dump:
```

**Comment by bshastry** [comment]
@chriseth Now this input causes a crash

```
{
  // Stores zero at u256(-1)
  mstore(0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff, 0x0)
  // Stores content of u256(-1) at 1
  mstore(1, mload(0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff))
}
```

The unoptimized trace is the following, the optimized trace is nothing (as expected). Looks like uninitialized memory in the interpreter to me but haven't taken a look yet.

```
Trace:
Memory dump:
     0: 0000000000000000000000000000000000000000000000000000000000000012
    20: 3300000000000000000000000000000000000000000000000000000000000000
Storage dump:
```

Ah, the `0x1233` is due to this

https://github.com/ethereum/solidity/blob/9a6357ab091c3331e7400bae887cd23540f07167/test/tools/yulInterpreter/EVMInstructionInterpreter.cpp#L254

**Comment by bshastry** [comment]
@chriseth Could you please rebase this PR onto develop?

**Comment by chriseth** [comment]
Rebased.

File: Changelog.md [hunk 1]
```diff
@@ -6,7 +6,7 @@ Language Features:
 
 Compiler Features:
  * Optimizer: Add rule that replaces the BYTE opcode by 0 if the first argument is larger than 31.
- * Yul Optimizer: Take side-effect-freeness of user-defined functions into account.
```
Comment (chriseth) [inline Changelog.md:10]: Should not be removed.

**Comment by bshastry** [comment]
@chriseth one more crashing test case

```
{
  mstore(not(0),1)
  extcodecopy(0,0,0,mload(not(0)))
}
```

is correctly optimized to

```
        {                                                             
            mstore(not(0), 1)
            extcodecopy(0, 0, 0, 1)
        }
```

but yul interpreter traces differ

```
Unoptimized Trace:
  EXTCODECOPY(0x01ffffffffffffffffffffffffffffffffffffffffffff, 0x1fffffffffffffffffffffffffffffffffffffffffffff, 0x01ffffffffffffffffffffffffffffffffffffffffffffff, 0)
Memory dump:
Storage dump:

Optimized Trace:
  EXTCODECOPY(0x01ffffffffffffffffffffffffffffffffffffffffffff, 0x1fffffffffffffffffffffffffffffffffffffffffffff, 0x01ffffffffffffffffffffffffffffffffffffffffffffff, 0x1fffffffffffffffffffffffffffffffffffffffffff)
Memory dump:
Storage dump:
```

I think this is because of my buggy PR #7330 where mload of an unaddressable address returns zero. Your new PR #7342 fixes this :-)

The trace from #7342 looks like this

```
Unoptimized Trace:
  EXTCODECOPY(0x01ffffffffffffffffffffffffffffffffffffffffffff, 0x1fffffffffffffffffffffffffffffffffffffffffffff, 0x01ffffffffffffffffffffffffffffffffffffffffffffff, 0x1fffffffffffffffffffffffffffffffffffffffffff)
Memory dump:
  10000000000000000000000000000000000000000000000000000000000000000: 0000000000000000001fffffffffffffffffffffffffffffffffffffffffff00
Storage dump:

Optimized Trace:
  EXTCODECOPY(0x01ffffffffffffffffffffffffffffffffffffffffffff, 0x1fffffffffffffffffffffffffffffffffffffffffffff, 0x01ffffffffffffffffffffffffffffffffffffffffffffff, 0x1fffffffffffffffffffffffffffffffffffffffffff)
Memory dump:
  10000000000000000000000000000000000000000000000000000000000000000: 0000000000000000001fffffffffffffffffffffffffffffffffffffffffff00
Storage dump:
```

I still need to understand what this new trace means.

File: libyul/optimiser/LoadResolver.h [hunk 2]
```diff
@@ -39,18 +41,29 @@ struct EVMDialect;
 class LoadResolver: public DataFlowAnalyzer
 {
 public:
+	/// Run the load resolver or the given complete AST.
```
Comment (bshastry) [inline libyul/optimiser/LoadResolver.h]: Type: Should be `Run the load resolver on the given complete AST`

Comment (bshastry) [reply libyul/optimiser/LoadResolver.h]: *Typo :laughing:

**Comment by chriseth** [comment]
Fixed.

The new trace is exactly the same, just that it does not stop at 2**256-1.

File: test/libyul/yulOptimizerTests/fullSuite/medium.yul [hunk 3]
```diff
@@ -21,11 +21,10 @@
...
-//         let _1 := 0x40
-//         mstore(_1, add(mload(_1), 0x20))
-//         let p := mload(_1)
-//         mstore(_1, add(p, _1))
-//         mstore(add(p, 96), 2)
-//         mstore(_1, 0x20)
+//         let _1 := mload(0x40)
+//         mstore(0x40, add(_1, 0x20))
+//         mstore(0x40, add(_1, 96))
+//         mstore(add(_1, 128), 2)
```
Comment (axic) [inline test/libyul/yulOptimizerTests/fullSuite/medium.yul:27]: Why did this became 128?

Comment (chriseth) [reply test/libyul/yulOptimizerTests/fullSuite/medium.yul:27]: The meaning of `_1` changed significantly, I think the code is still correct. Let me try `yulrun`.

Comment (chriseth) [reply test/libyul/yulOptimizerTests/fullSuite/medium.yul:27]: Yep, both versions store at 128.

Comment (ekpyron) [reply test/libyul/yulOptimizerTests/fullSuite/medium.yul:27]: It used to be ``add(p, 96)``, where ``p _:= mload(_1)`` - and there is ``add(mload(0x40), 0x20)`` stored at ``_1`` at that point - i.e. ``add(p, 96)`` is ``add(mload(0x40), 96+0x20 = 128)`` - and ``mload(0x40)`` is stored in ``_1`` in the new version - so fine!

**Review by ekpyron — APPROVED** [review]

**Pull request merged**